### PR TITLE
Mykola melnyk homework java 4

### PR DESCRIPTION
**Pull request by Mykola-Melnyk**

### CODE REVIEW

File: java/mykola-melnyk/src/com/zoolatech/lecture1/tasks/_1/Main.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package com.zoolatech.lecture1.tasks._1;
```
Comment (tromanyuk) [inline java/mykola-melnyk/src/com/zoolatech/lecture1/tasks/_1/Main.java:1]: please remove all files that are not related to the HW #4

File: zoola-academy-homework.iml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (tromanyuk) [inline zoola-academy-homework.iml:1]: iml file should be removed from the PR

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_1/Container.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    }
+
+    class ContainerUpdatesListener {
+        private int firstValue;
+
+        public ContainerUpdatesListener() {
+            Container.this.value = firstValue;
+        }
+        public boolean hasValueChanged() {
+            if (Container.this.value == firstValue) {
```
Comment (tromanyuk) [inline java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_1/Container.java:25]: `==` already returns a boolean value, you can simplify your code by directly returning it, without the if condition

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_2/BinarySearchTree.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
+package com.zoolatech.lecture4.tasks._2;
+
+public class BinarySearchTree {
+    private Node rootNode;
+    public BinarySearchTree () {
+        rootNode = null;
```
Comment (tromanyuk) [inline java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_2/BinarySearchTree.java:6]: the default value of rootNode is null, you can remove both the constructor and initialization statement

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_2/BinarySearchTree.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    private static class Node {
+        private final int value;
+        private Node leftChild;
+        private Node rightChild;
+        public Node (int value) {
+            this.value = value;
+            leftChild = null;
```
Comment (tromanyuk) [inline java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_2/BinarySearchTree.java:15]: same: leftNode and rightNode initialization statements can be removed

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_2/BinarySearchTree.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        } else if (current.value > value) {
+            if (current.leftChild == null) {
+                current.leftChild = new Node(value);
+            } else {
+                addRecursive(current.leftChild, value);
+            }
+        }
+    }
+
+    public void add(int value) {
```
Comment (tromanyuk) [inline java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_2/BinarySearchTree.java:36]: methods should be defined in the usage order: method `add` uses the method `addRecursive`, so it's better to place `addRecursive` after `add`

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_3/Task3.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+public class Task3 {
+
+    public static void modifyString(String string, StringModifier stringModifier) {
+        System.out.println(stringModifier.modify(string));
+    }
+
+    public static void main(String[] args) {
+        StringModifier toUpperCase = string -> {
+            return "modified: " + string.toUpperCase();
```
Comment (tromanyuk) [inline java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_3/Task3.java:22]: since your lambda expression has only a single line, you don't need to use the return statement and braces here

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_3/Task3.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+        StringModifier toLowerCase = string -> {
+            return "modified: " + string.toLowerCase();
+        };
+
+        modifyString("Hello!", toUpperCase);
+        modifyString("WHAZZZUUUPPP???", toLowerCase);
+
+        modifyString("howdy", String::toUpperCase);
+        modifyString("BYOB is for BOOZE", String::toLowerCase);
+
```
Comment (tromanyuk) [inline java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_3/Task3.java:33]: please pay attention to extra new lines you're adding

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_4/WeightUnits.java [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+    public String getAbbreviation() {
+        return abbreviation;
+    }
+
+    public double convertFrom(double value, WeightUnits sourceUnit) {
+        double result = 0;
+        switch (this) {
+            case GRAM -> {
+                result = switch (sourceUnit) {
```
Comment (tromanyuk) [inline java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_4/WeightUnits.java:19]: you can return the value directly from the switch expression, without storing it to a local variable

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_1/Container.java [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+    public void setValue(int value) {
+        this.value = value;
+    }
+
+    class ContainerUpdatesListener {
+        private int firstValue;
+
+        public ContainerUpdatesListener() {
+            Container.this.value = firstValue;
```
Comment (tromanyuk) [inline java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_1/Container.java:22]: it should be opposite: you need to assign the `value` to the `firstValue`; note: since variable names are different you don't need to use a container's class name to access its variable

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_4/WeightUnits.java [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+                case OUNCE -> value / 35270;
+                case POUND -> value / 2205;
+                case TON -> value;
+                };
+            }
+        }
+        return result;
+    }
+
+    public WeightUnits getUnit(String name) {
```
Comment (tromanyuk) [inline java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_4/WeightUnits.java:67]: the method can be static

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_4/WeightUnits.java [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+        WeightUnits unit = null;
+        switch (name.toUpperCase()) {
```
Comment (tromanyuk) [inline java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_4/WeightUnits.java:69]: you can use the switch expression and return the value directly

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_5/Task5.java [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+    public static void main(String[] args) {
+        Scanner in = new Scanner(System.in);
+        System.out.println("Enter first number:");
+        BigDecimal firstnum = in.nextBigDecimal();
+        System.out.println("Enter operator (+, -, *, /, %):");
+        char operator = in.next().charAt(0);
+        System.out.println("Enter second number:");
+        BigDecimal secondnum = in.nextBigDecimal();
+        BigDecimal result = null;
+        switch (operator) {
```
Comment (tromanyuk) [inline java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_5/Task5.java:21]: same: you can use the switch expression here and assign the value directly to the variable

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_6/StringAnalyze.java [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+ * 	“8 letters”
+ * 	“2 digits”
+ * 	“3 spaces”
+ * 	“2 other characters”
+ */
+
+public class StringAnalyze {
+    public static void analyze(String input) {
+        int length = input.length();
+        char[] chars = new char[length];
```
Comment (tromanyuk) [inline java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_6/StringAnalyze.java:18]: there's a simple method how to get chars: input.toCharArray()

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_6/StringAnalyze.java [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+        input.getChars(0, length, chars, 0);
+        System.out.println(Arrays.toString(chars) + "\n" + "Char #: " + length);
+        int letterCounter = 0;
+        int digitCounter = 0;
+        int spaceCounter = 0;
+        int otherCounter = 0;
+        for (char e : chars) {
```
Comment (tromanyuk) [inline java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_6/StringAnalyze.java:25]: try to find a better naming for the variable `e`

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_7/DigitToString.java [hunk 16]
```diff
@@ -0,0 +1,32 @@
+package com.zoolatech.lecture4.tasks._7;
+
+import java.util.Arrays;
+
+public class DigitToString {
+
+    public static String convertToWord(char digit) {
+        String word = null;
+        switch (digit) {
```
Comment (tromanyuk) [inline java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_7/DigitToString.java:9]: same: you can use the switch expression here

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_7/DigitToString.java [hunk 17]
```diff
@@ -0,0 +1,32 @@
+package com.zoolatech.lecture4.tasks._7;
+
+import java.util.Arrays;
+
+public class DigitToString {
+
+    public static String convertToWord(char digit) {
```
Comment (tromanyuk) [inline java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_7/DigitToString.java:7]: I think this method should be called `digitToString`, and the next one - `convertToWord`

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_7/DigitToString.java [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+            case '9' -> word = "nine";
+            case '0' -> word = "zero";
+            default -> word = "Not a Digit!";
+        }
+        return word;
+    }
+    public static void digitToString (String stringOfDigits) {
+        int length = stringOfDigits.length();
+        char[] chars = new char[length];
+        stringOfDigits.getChars(0, length, chars, 0);
```
Comment (tromanyuk) [inline java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_7/DigitToString.java:27]: `stringOfDigits.toCharArray`

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_7/DigitToString.java [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+        for (Character e : chars) {
```
Comment (tromanyuk) [inline java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_7/DigitToString.java:28]: please find a better naming for the variable `e`

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_7/DigitToString.java [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+            System.out.print(e + "(" + convertToWord(e) + ")");
```
Comment (tromanyuk) [inline java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_7/DigitToString.java:29]: methods should be defined in the usage order: it's better to define `convertToWord` after this method

File: java/mykola-melnyk/src/com/zoolatech/lecture1/tasks/_1/Main.java (hunk 1 above)
Comment (Mykola-Melnyk) [reply java/mykola-melnyk/src/com/zoolatech/lecture1/tasks/_1/Main.java:1]: Can you please specify which files exactly do I need to remove?

File: zoola-academy-homework.iml (hunk 2 above)
Comment (Mykola-Melnyk) [reply zoola-academy-homework.iml:1]: Done

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_1/Container.java (hunk 3 above)
Comment (Mykola-Melnyk) [reply java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_1/Container.java:25]: Done

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_2/BinarySearchTree.java (hunk 4 above)
Comment (Mykola-Melnyk) [reply java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_2/BinarySearchTree.java:6]: Done

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_2/BinarySearchTree.java (hunk 5 above)
Comment (Mykola-Melnyk) [reply java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_2/BinarySearchTree.java:15]: Done

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_2/BinarySearchTree.java (hunk 6 above)
Comment (Mykola-Melnyk) [reply java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_2/BinarySearchTree.java:36]: Done

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_3/Task3.java (hunk 7 above)
Comment (Mykola-Melnyk) [reply java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_3/Task3.java:22]: Done

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_3/Task3.java (hunk 8 above)
Comment (Mykola-Melnyk) [reply java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_3/Task3.java:33]: Done

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_4/WeightUnits.java (hunk 9 above)
Comment (Mykola-Melnyk) [reply java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_4/WeightUnits.java:19]: Done

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_1/Container.java (hunk 10 above)
Comment (Mykola-Melnyk) [reply java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_1/Container.java:22]: Done

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_4/WeightUnits.java (hunk 11 above)
Comment (Mykola-Melnyk) [reply java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_4/WeightUnits.java:67]: Done

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_4/WeightUnits.java (hunk 12 above)
Comment (Mykola-Melnyk) [reply java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_4/WeightUnits.java:69]: Done

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_5/Task5.java (hunk 13 above)
Comment (Mykola-Melnyk) [reply java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_5/Task5.java:21]: Done

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_6/StringAnalyze.java (hunk 14 above)
Comment (Mykola-Melnyk) [reply java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_6/StringAnalyze.java:18]: Done

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_6/StringAnalyze.java (hunk 15 above)
Comment (Mykola-Melnyk) [reply java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_6/StringAnalyze.java:25]: Done

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_7/DigitToString.java (hunk 16 above)
Comment (Mykola-Melnyk) [reply java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_7/DigitToString.java:9]: Done

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_7/DigitToString.java (hunk 20 above)
Comment (Mykola-Melnyk) [reply java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_7/DigitToString.java:29]: Done

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_7/DigitToString.java (hunk 19 above)
Comment (Mykola-Melnyk) [reply java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_7/DigitToString.java:28]: Done

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_7/DigitToString.java (hunk 17 above)
Comment (Mykola-Melnyk) [reply java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_7/DigitToString.java:7]: Done

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_7/DigitToString.java (hunk 18 above)
Comment (Mykola-Melnyk) [reply java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_7/DigitToString.java:27]: Done

File: java/mykola-melnyk/src/com/zoolatech/lecture1/tasks/_1/Main.java (hunk 1 above)
Comment (tromanyuk) [reply java/mykola-melnyk/src/com/zoolatech/lecture1/tasks/_1/Main.java]: all files from the folder `java/mykola-melnyk/src/com/zoolatech/lecture3
Screenshot 2022-08-22 at 17 03 58

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_4/WeightUnits.java (hunk 12 above)
Comment (tromanyuk) [reply java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_4/WeightUnits.java:67]:
```
return switch(name.toUpperCase()) {
    case "GRAM" -> WeightUnits.GRAM;
    case "KILOGRAM" ->  WeightUnits.KILOGRAM;
    ...
}
```

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_7/DigitToString.java [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+public class DigitToString {
+
+    public static void convertToWord(String stringOfDigits) {
+        char[] chars = stringOfDigits.toCharArray();
+        for (Character character : chars) {
+            System.out.print(character + "(" + digitToString(character) + ")");
+        }
+    }
+    public static String digitToString(char digit) {
+        switch (digit) {
```
Comment (tromanyuk) [inline java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_7/DigitToString.java:12]:
```
return switch(digit) {
    case '1' -> "one";
    case '2' -> "two";
   ...
}
```

File: java/mykola-melnyk/src/com/zoolatech/lecture1/tasks/_1/Main.java (hunk 1 above)
Comment (Mykola-Melnyk) [reply java/mykola-melnyk/src/com/zoolatech/lecture1/tasks/_1/Main.java]: Deleted all the contents of the lecture 3 folder. But I must have misunderstood something. What about the lectures 1&2? Do I have to remove the contents of thhat folders too? I don't get the idea, sorry. Can you please explain the logic behind this? Thank you :)

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_4/WeightUnits.java (hunk 12 above)
Comment (Mykola-Melnyk) [reply java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_4/WeightUnits.java:67]: Thank you, corrected.

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_7/DigitToString.java (hunk 21 above)
Comment (Mykola-Melnyk) [reply java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_7/DigitToString.java:12]: Thank you, corrected

File: java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_5/Task5.java (hunk 13 above)
Comment (Mykola-Melnyk) [reply java/mykola-melnyk/src/com/zoolatech/lecture4/tasks/_5/Task5.java:20]: Rewrote this, thanks!

File: java/mykola-melnyk/src/com/zoolatech/lecture1/tasks/_1/Main.java (hunk 1 above)
Comment (tromanyuk) [reply java/mykola-melnyk/src/com/zoolatech/lecture1/tasks/_1/Main.java]: When you create a new branch from the current branch, the new branch will contain all commits from the current one. I think when you created the branch for the current MR ("mykola-melnyk-java-4"), you did it from the branch with your 3rd homework ("mykola-melnyk-java-3"). So, now the branch for this MR contains all changes from both homeworks. To prevent such issues you need to create new branches for your homeworks from the master branch only.

The simplest way to resolve this issue might be to create a new branch from the master branch, use [git cherry pick](https://git-scm.com/docs/git-cherry-pick) tool to copy commits from the old branch, and remove the old branch.

To simplify the process you can keep the files from other homeworks.

**Review by tromanyuk — APPROVED** [review]

File: java/mykola-melnyk/src/com/zoolatech/lecture1/tasks/_1/Main.java (hunk 1 above)
Comment (Mykola-Melnyk) [reply java/mykola-melnyk/src/com/zoolatech/lecture1/tasks/_1/Main.java]: Thank you so much for your detailed answer!

**Comment by Mykola-Melnyk** [comment]
There was another .gitignore conflict, I just accepted the variant from main. Is that correct?

**Review by tromanyuk — APPROVED** [review]
